### DISCUSSION TITLE
canonical marshaling

### DISCUSSION THREAD

**Discussion by missinglink** [whosonfirst/go-whosonfirst-export#19] [General]
I would help develop & maintain a command which is responsible for 'marshaling' WOF features into a 'canonical byte format'.
A stable byte encoding is essential for tools like `diff`, `git` etc. to work as expected and makes PR review easier.

#### A canonical byte format

This 'canonical byte format' initially seems irrelevant, as it's not required for moving data *from* the repos/bundles/sqlites *into* a foreign application for consumption (such as in the case of Pelias and other consumers of the data), or passing the data between applications.

However, when trying to move data *back* into the repos via pull requests, the data often differs in byte encoding, things like numbers/floats, UTF equivalence, property ordering, newlines/lastline, etc. become a blocker for contribution and end up muddying the diffs.

The behaviours of these [marshalling quirks](https://github.com/whosonfirst/py-mapzen-whosonfirst-export/issues/28) from various languages are particularly difficult to mitigate without post-processing the output string using complex regular expressions, and would likely never reach perfect byte-for-byte compatibility with those produced by the official tools.

This is where I feel a tool equivalent to `go fmt` would be particularly useful, it would allow external contributors to move the data from git, into their application memory, then back to git without introducing marshalling quirks along the way.

#### Existing tools

**exportify**

The existing `exportify` tools perform a variety of tasks which prepare the feature for merging into git, the final stage of which is marshaling the feature to disk (as discussed above).

I would like to have this functionality decoupled from the other tasks it performs.

This would allow external developers to make arbitrary changes to features and write them back to a git fork for diff/review without also having their changes conflated with the other tasks performed by `exportify`.

The `exportify` command is also not appropriate for tasks where you would like to normalize the byte encoding but without mutating the structure of the feature.

An example of this is when generating bundles in nodejs, the data is unmarshalled into memory using `JSON.parse()` to find the ID and other relevant information (for meta files, spr table etc.) and then written to the bundle/database.

If the application keeps two copies of each feature in memory, one as bytes and one as an in-memory structure then it can write the original bytes into the bundle, but this is not possible for any program written using a streaming interface as the original byte encoding is lost.

Using `exportify` for this task could cause a mutation of the data structure, which is undesirable in this context.

**go-whosonfirst-format**

There is also a [go-whosonfirst-format](https://github.com/whosonfirst/go-whosonfirst-format) tool which I wasn't aware of until today which does a lot of what I am looking for.

If we can use this to generate 'canonical marshaling' then that's great, although I would remove any configurability (such as `pretty` and `newline`) which would allow the library to produce anything other than a canonical format.

**what's currently being used?**

It's not really clear which marshaling format is currently in use in the `whosonfirst-data` repos, I suspect it's a mix of different algorithms in different languages over the years, although this is somewhat of a moot point, once we have this tool we can simply reformat everything into the canonical encoding.

#### Summary

Having a single tool which 'does one thing and one thing well' would mean that external developers could write their own tooling in their own preferred language and have a single lean dependency for preparing their contributions for review.

**Comment by missinglink** [comment]
Here's a few things I'd like to see in this tool:

- it should normalize how properties are ordered and encoded.
- it must not modify the meaning of the data (ie. it does not add/remove properties) *this might need a clearer/stricter definition*
- it must be [Idempotent](https://en.wikipedia.org/wiki/Idempotence).
- it should be strongly opinionated and non configurable.
- it should be able to operate on a single file, a collection of files, and operate via stdio.
- it should be compiled into a single binary which can be easily distributed.
- it should have no external dependencies.
- it should be well tested and easy to contribute patches to.
- it should be available as *either* a library *or* a binary.

**Comment by thisisaaronland** [comment]
@missinglink Thanks for this. Today is a little busy for me so I may not be able to offer useful comments right away.

Paging @tomtaylor and @vicchi to the formatting and validation courtesy telephones.

**Comment by thisisaaronland** [comment]
Briefly, because I have a spare moment. There has been an ongoing effort to split the various pieces of "exporting" a WOF document in to discrete packages. They are:

* [go-whosonfirst-format](https://github.com/whosonfirst/go-whosonfirst-format) - This has no opinion about the _values_ contained in a GeoJSON record. It only cares about how those bytes are formatted. The rules are: All properties are sorted alphabetically and are indented. The exception to this rule are geometries which are _not_ indented and included at the end of a GeoJSON feature record. The rules about indenting, if any, remain in flux. That's it.

* [go-whosonfirst-export](https://github.com/whosonfirst/go-whosonfirst-export) - This is where WOF has opinions about data to export, adding or modifying data as necessary and throwing errors where conditions are invalid or can not be met. It is also an interface for _other_ project-specific exporters (for example [SFO Museum](https://github.com/sfomuseum/go-sfomuseum-export) and a way to write code such there are only `export.Exporter` instances and the code doesn't know (or care) which project it is exporting to.

* [go-whosonfirst-id](https://github.com/whosonfirst/go-whosonfirst-id) - This is wrapper and utility code for generating WOF IDs (aka: Brooklyn Integers (BI)). As of this writing it is wrapped (tangled?) up in two other abstraction layers: One for minting UIDs in general and another to provide a proxy and a cache pool in front of BI for processes that need to mint a lot of IDs. The BI piece itself is wrapped up in an abstraction layer for artisanal integers, generally. The whole thing borders on the absurd. I am not wed to any particular piece of it but experience has demonstrated that people often have needs that are more complicated than simply "making an HTTP request to BI".
 
* [go-whosonfirst-validate](https://github.com/whosonfirst/go-whosonfirst-validate) - This one hasn't been updated in a long time. It remains more aspirational than useful but it should be a place to test and validate WOF features independent of exporting them. This is where all of @vicchi 's work around (JSON) schemas and validation would intersect.

* [go-writer](https://github.com/whosonfirst/go-writer) (and friends) - These are the packages that write bytes to a target (a writer). They only care about moving bytes from (a) to (b) and are in no way WOF/GeoJSON specific. 

* [go-whosonfirst-exportify](https://github.com/whosonfirst/go-whosonfirst-exportify) - These are meant to be WOF-specific CLI tools and, importantly, this is the package that doesn't really concern itself with the size or volume of dependencies. That is there are tools for performing specific functions that depend on packages that don't have any business being requirements for the packages listed above.

**Reply by missinglink** [reply]
Cool, yeah it sounds like `go-whosonfirst-format` is where I want to focus my attention.

It seems as though the wind is blowing in this direction as far as it becoming the new officially sanctioned/preferred marshaling tool (including its dependants)?

**Comment by thisisaaronland** [comment]
"...the new officially sanctioned/preferred marshaling tool"

To the extent that @stepps00 still does most of his work in Python and needs to be able to export records as/from library code there should always be an equivalent Python port.

For general purpose "export one or more static files" use cases I would prefer that the Go version be the official version because we remove the need to ask people of going down the rabbit hole of Python and its twisty maze of dependencies (not least of which is gdal).

**Reply by missinglink** [reply]
Right, I guess we could always wire the python tooling up to run `go-whosonfirst-format`.

**Comment by thisisaaronland** [comment]
I think the only outstanding difference between the Python and Go libraries is the question of indenting and since the bulk of the records have already been formatted/indented by the Python code we should work to make sure that `go-whosonfirst-format` does the same thing.

**Reply by tomtaylor** [reply]
So, from memory, when I did the first version of `go-whosonfirst-format` I treated the Python library as canonical. However, there's a bunch of potentially minor differences around trailing whitespace, for example. 

Maybe we could have a shared test suite for both libraries to make sure they align?

**Reply by tomtaylor** [reply]
See previous chat: https://github.com/whosonfirst/go-whosonfirst-format/issues/1

In retrospect, looking at the commits, it looks like I couldn't quite face Python's inconsistent indentation spec, so went with something a bit more straightforward.

**Reply by missinglink** [reply]
> Maybe we could have a shared test suite for both libraries to make sure they align?

I suspect that getting two different languages to serialize JSON in exactly the same way is the ultimate yak shaving exercise 😆 🐃 

It goes beyond indentation, sorting and non-printing characters right through to things like [floating point numbers](https://stackoverflow.com/questions/58411805/floating-point-accuracy-with-different-languages), [unicode normalization/equivalence](https://en.wikipedia.org/wiki/Unicode_equivalence), [numeric types](https://github.com/whosonfirst/py-mapzen-whosonfirst-export/issues/28) inc. scientific notation, ...∞.

**Reply by tomtaylor** [reply]
Of course, but _if_ the goal is to have mostly byte for byte compatibility to reduce diff churn there’s probably a few (10-30?) scenarios we need to cover off. If that’s not the goal, then a few bullet points for the intended spec should suffice.

**Reply by tomtaylor** [reply]
I also agree that using the Go formatter from Python would be simpler.

**Reply by thisisaaronland** [reply]
Has anyone done any "calling Go from Python" work? At a glace it looks mostly like building a binary with specific C variables exposed that Python call read. If we could simply expose an `export` method to Python along the lines of the WASM work that would be great:

* https://github.com/whosonfirst/go-whosonfirst-export-wasm

**Reply by missinglink** [reply]
I'm definitely advocating something which guarantees *exact* byte for byte compatibility rather than being *mostly* compatible.

There's this great quote from Rob Pike:
> Gofmt's style is nobody's favourite, but gofmt is everybody's favourite.

It's true, I can't say I would choose the code formatting conventions that they did, but there's one single way of formatting Go code and that's actually really awesome and means everyone can get on with writing code instead of arguing about whether lines should be indented with spaces or tabs 😆 

And that's the real value of this thing, it's that everyone can use WOF in their own code and in their own language and when it comes time to merge contributors back the diffs all look like they're authored by the same person.

**Comment by thisisaaronland** [comment]
Picking a random (read: the first) record from `whosonfirst-data-admin-us` it appears the indenting is "2 spaces" which is, I believe, the default for Python. @tomtaylor did you see otherwise.

If that is the case, then I would suggest that be the rule for all the other things. Basically it's simple and dumb and easy to explain and implement.

**Reply by tomtaylor** [reply]
I have seen a bunch of other stuff, yes. I’ve seen files with mixed 2/4 space indentation, trailing whitespace after the comma, and inconsistent key ordering. Just a few that spring to mind!

I agree 2 space indentation is the norm, and moving towards that has the minimum churn.

**Reply by thisisaaronland** [reply]
I am happy to say:

* Order all keys alphabetically, with the exception of the top-level `geometry` key (meaning keys with in the `geometry` dictionary are ordered alphabetically)
* 2-space indentation
* No trailing-whitespace after commas
* No scientific notation
* Coordinates are capped at 13 decimal points 👈 @nvkelso is that right?

Unicode normalization should probably be whatever Go does since Go, collectively, has more experience with the issues and spends more time sweating the details. Which is a bummer if you're not using Go and another reason to about building/exposing something that can be invoked from the many C-bridges that exist.

As to type values that's where the `whosonfirst-properties` specifications and/or @vicchi 's JSON schema work (which is derived from the properties spec) comes in to play and should specific to the `go-whosonfirst-export` and `go-whosonfirst-validate` packages. Those things aren't about formatting.

If we figure out a way to expose the Go functionality to (some other language) for the export piece then it stands to reason the same can be done for a validation piece.

Everyone: What else am I missing?

**Reply by tomtaylor** [reply]
Off the top of my head:

- Array/map formatting behaviour (new lines for every element or only if the line needs wrapping?)
- Trailing new line to end the file

**Reply by thisisaaronland** [reply]
I would say:

* New lines for every array element because then there is no question about whether or not to add them.
* No trailing new line at the end of the file but I know there are reasons why this is necessary sometimes. I don't remember what any of those reasons though...

**Reply by tomtaylor** [reply]
At the moment the Go formatter does add a new line at the end of every file because… that’s what Python does, I think.

**Reply by missinglink** [reply]
A couple other things I can think of off the top of my head:
- Languages that convert `1.0` to `1` (such as JS) should have them cast back to the correct type [according to the spec](https://github.com/whosonfirst/whosonfirst-properties/blob/main/properties/lbl/max_zoom.json#L6).
- For unicode I would suggest normalizing to NFKC and then stripping any extra combining chars, see https://github.com/pelias/pelias/issues/829#issuecomment-542614645

**Reply by thisisaaronland** [reply]
@missinglink  Both of these seem out of scope for _formatting_ piece. It's understood that the conversation is all over the map right now so I just want to make sure that suggestions and concerns get put in their relevant "bucket" (or package).

The first - trying to recast types to meet the properties spec - is something that should be handled by the code in `go-whosonfirst-export/properties`. That's the place where WOF asserts an editorial opinion about things.

The second I need to think about. It has the effect of silently rewriting other people's data which feels like a bad idea.

**Reply by missinglink** [reply]
> * Coordinates are capped at 13 decimal points

There's a great [stackoverflow post](https://gis.stackexchange.com/a/8674) about this, I usually truncate to 6 (rounding down), but I can see how 7 might be chosen instead.

Beyond 11mm I feel like it's more likely to be error introduced by the technology used to measure it than super high fidelity.

**Reply by missinglink** [reply]
> Both of these seem out of scope for _formatting_ piece

Agh yeah I can see that, unfortunately the first has [blocked contributions](https://github.com/whosonfirst-data/whosonfirst-data-admin-ca/pull/23#pullrequestreview-444860428) from me in the past and it would be required for me to be able to get PRs generated in nodejs merged back into WOF.

**Reply by missinglink** [reply]
It's not unique to nodejs, features passed through JQ [suffer the same problem](https://github.com/whosonfirst/py-mapzen-whosonfirst-export/issues/28) and it's written in C.

Numbers are a tricky one, you could argue that while `1.0` and `1` are numerically equivalent, the trailing `.0` is just formatting 🤷‍♂️

**Reply by thisisaaronland** [reply]
Per the number-formatting (which is really an int versus float type) issue the simplest-dumbest-quickest way to handle that in the Go code would be to add a new method in `go-whosonfirst-export/properties` and invoke it from [prepareWithoutTimestamps](https://github.com/whosonfirst/go-whosonfirst-export/blob/d6beed9f9e01ce3fcbe2388b7a17411c64c92877/export.go#L107).

In the beginning that might mean a lot of boring if-this-then-that code but at least it would be encapsulated and could eventually be replaced by code that reads definitions from [whosonfirst-properties](https://github.com/whosonfirst/whosonfirst-properties) whether that's a JSON schema or something else.

The other thing the `1.0` versus `1` problems points to is the idea that some data is considered "invalid" but not strict in that there might be code that will fix things for you.

**Comment by nvkelso** [comment]
> Coordinates are capped at 13 decimal points 👈 @nvkelso is that right?

I think we've been doing 6 decimal places more recently, which has been fine. Originally we'd errored on side of preserving more-or-less the source data precision on import... but practically speaking 6 is fine and what we've been doing the last couple years.

**Comment by thisisaaronland** [comment]
Here is a thing I had forgotten about:

* https://github.com/whosonfirst/go-whosonfirst-properties

This was originally written to _generate_ properties files and I have done only enough work to bring the package up to date with Go modules and to use the `go-whosonfirst-iterate` packages in place of the deprecated `go-whosonfirst-index` packages.

This package could be updated easily to contain a lookup table of property definitions, whether its pre-compiled along the lines of the `go-whosonfirst-placetypes` package or generated at runtime. That could then be used by the `-export` and/or `-validate` packages for sanity checking values and types.

**Reply by missinglink** [reply]
Yeah, a simple solution would be to just [embed](https://pkg.go.dev/embed) the JSON files in the binary.

**Comment by missinglink** [comment]
Hi all, I had a think about a clearer definition of the scope of this tool over the weekend.

What I think it boils down to is the difference between 'equivalent' and 'equal'.
> "Given ambiguous equivalence, `woffmt` should select a canonical form and enforce that" 

Some examples:

```
# spacing equivalent but not equal
{"wof:foo":"a"}
{ "wof:foo": "a" }

# objects equivalent but not equal
{ "wof:foo": "a", "wof:bar": "b" }
{ "wof:bar": "b", "wof:foo": "a" }

# numbers equivalent but not equal
{ "wof:foo": 1.0 }
{ "wof:foo": 1 }
{ "wof:foo": 1.0e0 }

# strings equivalent but not equal
{ "wof:foo": "Köln" }
{ "wof:foo": "Köln" }
{ "wof:foo": "\u004b\u00f6\u006c\u006e" }

# but not transliterations!
{ "wof:foo": "Koeln" }

# empty values vs non-existence
{ "wof:foo": [] }
{}
```

**Comment by tomtaylor** [comment]
FYI, I did a small amount of work on https://github.com/whosonfirst/go-whosonfirst-format to fix a few loose ends and expose a binary you can pipe through for formatting.

**Comment by straup** [comment]
Adding to the list of packages in the layer cake:

* [go-whosonfirst-writer](https://github.com/whosonfirst/go-whosonfirst-writer) - This is a utility/helper package that sits on top of the other packages (described above) and is meant to encapsulate (read: make less-typing) common export/format/write scenarios. It isn't meant, and shouldn't be used, for edge-cases or complex publishing needs.

* [go-whosonfirst-reader](https://github.com/whosonfirst/go-whosonfirst-reader) and [go-reader](https://github.com/whosonfirst/go-reader) - These are the read-equivalent of the `go*-writer` packages.

* [go-whosonfirst-geojson-v2](https://github.com/whosonfirst/go-whosonfirst-geojson-v2) - This one can't be retired soon enough but it's going to take a little while because it's used everywhere and a replacement interface/API hasn't been completely sorted out yet. Basically, marshaling/unmarshaling and passing around WOF-derived GeoJSON instances has proven to be unnecessary. I think a better approach would be to settle on passing around `[]byte` arrays (typically for processing by `tidwall/gjson` or `tidwall/sjson`) and when an actual GeoJSON instance is needed just using `paulmach/orb/geojson`. The things that are of most value in this package are the methods defined in the `properties` subpackage and those could all be updated accordingly.The `-v2` part reflects the fact that this package was first released before Go had figured out a common versioning system. My preference would be to replace this package with a generic `go-whosonfirst-record` (or maybe `-feature`) package that implements all of the above.

**Reply by thisisaaronland** [reply]
Of course...

* [go-whosonfirst-feature](https://github.com/whosonfirst/go-whosonfirst-feature)

**Comment by missinglink** [comment]
Good news, after looking at this some more I have to admit I was wrong about "getting two different languages to serialize JSON in exactly the same way being the ultimate yak shaving exercise" 😆 

It's true that standard library functions (such as `json.Marshal()` in Go and `JSON.stringify()` in Javascript) make this nay impossible but what I didn't realize was just how simple a reliable JSON serialiser is to write yourself.

So I've changed by position on this now and I'd be happy to help write a spec which can be implemented in both Go and JS (to begin with).

It seems that the `whosonfirst-properties` repo is the ideal place to store the 'serialization rules', WOF has a few quirky things, such as enforcing lexographical sorting of object keys *except* for the root.

I found the easiest way to implement a custom serialiser was to have 'standard rules' (ie. about indentation, key ordering, delimiter spacing, etc.) and then allowing for a deviation from that standard based on specific elements, using some form of JPath syntax seems to be ideal.

So in effect what I'm talking about for the ruleset would be similar to CSS, where you have a base styling and then you can specify individual rules by using a selector syntax to target elements in the DOM tree.

An example of a 'canonical encoding ruleset' which I found to produce consistent and aesthetically pleasing results (written as CSS as a familiar example, but not necessarily suggesting that format!):

```go
* {
	--newline: "\n";
	--array-open: "[";
	--array-close: "]";
	--array-delim: ", ";
        --object-indent: "  ";
        --object-open: "{";
        --object-close: "}";
        --object-delim: ",";
        --object-sep: ": ";
        --object-key-order: "";
	--float-min-precision: -1;
	--float-max-precision: -1;
	--string-unicode-normalization: "";
}

object.root {
	--object-key-order: "id", "type", "properties", "bbox", "geometry";
}

object.geometry {
	--object-key-order: "type", "coordinates";
}

.properties .lbl:max_zoom {
	--float-min-precision: 1;
	--float-max-precision: 1;
}

.geometry .coordinates {
	--float-min-precision: 0;
	--float-max-precision: 6;
}
```

Given the above I'm pretty confident that we can produce the same serialisation in Python, Go and the rest.

**Comment by missinglink** [comment]
Here's an example diff using the rules above, there are some aspects (such as the arrays being contracted) which may not be everyone's cup 'o tea but we can discuss those when defining the actual spec.

https://www.diffchecker.com/p5xrS9c4

**Reply by nvkelso** [reply]
Looking good!

Nit: when a list has contents then each element should be on it's own line (but if no contents then single line with `"*": []` ... else it's hard in later PRs to know which specific value is changing. This is true for `"name:*"` properties, and other `"wof:*"` properties.
